testsuite/tests/typing-modes/val_modalities.ml

```diff
@@ -73,88 +73,121 @@ module Module_type_of_comonadic = struct
     module M = struct
         let x @ portable = fun x -> x
     end
-    (* for comonadic axes, we default to id = meet_with_max, which is the
-    weakest. *)
+    (* for comonadic axes, we default to meet_with_min, which is the strongest.
+    *)
     module M' : module type of M = struct
         let x @ portable = fun x -> x
     end
     let _ = portable_use M.x (* The original inferred modality is zapped *)
 end
 [%%expect{|
-Line 10, characters 25-28:
-10 |     let _ = portable_use M.x (* The original inferred modality is zapped *)
-                              ^^^
-Error: This value is "nonportable" but expected to be "portable".
+module Module_type_of_comonadic :
+  sig
+    module M : sig val x : 'a -> 'a @@ stateless end
+    module M' : sig val x : 'a -> 'a @@ stateless end
+  end
+|}]
+
+(* zapping behavior can cause type error that shouldn't happen in upstream ocaml *)
+module Module_type_of_error = struct
+  module M = struct
+    let x = fun x -> x
+  end
+
+  module M' : module type of M = struct
+    let y = ref 42
+    let x = fun x -> ignore !y; x
+  end
+end
+[%%expect{|
+Lines 6-9, characters 33-5:
+6 | .................................struct
+7 |     let y = ref 42
+8 |     let x = fun x -> ignore !y; x
+9 |   end
+Error: Signature mismatch:
+       Modules do not match:
+         sig val y : int ref val x : 'a -> 'a end
+       is not included in
+         sig val x : 'a -> 'a @@ stateless end
+       Values do not match:
+         val x : 'a -> 'a
+       is not included in
+         val x : 'a -> 'a @@ stateless
+       The second is portable and the first is nonportable.
 |}]
 
 module Module_type_of_monadic = struct
     module M = struct
-        let x @ uncontended = ref "hello"
+        let x : string ref @ uncontended = ref "hello"
```

Comment:
I was intrigued by the need to add this and so checked what would happen without. There is a different output in `-principal` mode. Do we know why? Is the deviation of principal mode here tracked by a CR or ticket? I think I'd prefer not adding the type here so that we see the discrepancy (so we can fix it some day).

Reply:
Mode crossing is guarded by a `is_principal` check, which means in principal mode some mode crossing are no longer availab.e

Reply:
Is it just different printing? I seem to remember seeing different inferred modalities. I think that, even in principal mode, there should be no difference here, because the type `string ref` is principally determined (by generalization) before any zapping happens.

To be clear, I'm fine with not fixing this now, but I think adding the type signatures has hidden (rather than delayed) the bug.

Reply:
I don't remember the details, but the mode crossing in question is probably 1. argument to `ref` is mode crossed 2. result of `ref` is mode crossed. Both are before the generalization.

Reply:
I don’t want this patch caught here. Maybe just remove the type signature, accept the strange `Principal` output, and leave a CR to return later?

Reply:
I still think leaving a CR would be good. The CR makes it more likely we'll actually get back here someday. Include a link to this discussion in the CR, so we can remember this context later. Thanks.